### ISSUE TITLE
Unable to parse time range. Please put it as a range of time specs Tcl's \[clock sc...

### ISSUE THREAD

**Issue by Rendiarca** [ianka/xfs_undelete#45]
`./xfs_undelete -t 2024-01-03 -r 'zip/*' /dev/mapper/centos-home
invalid command name "lmap"
    while executing
"lmap t $times {
                if {[catch {clock scan $t} t]} {
                        puts stderr "Unable to parse time range. Please put it as a range of time specs Tcl's \[clock sc..."
    (procedure "parseTimerange" line 13)
    invoked from within
"parseTimerange [dict get $::parameters t]"
    invoked from within
"set ctimes [parseTimerange [dict get $::parameters t]]"
    (file "./xfs_undelete" line 460)
`

always error when executed

**Comment by Rendiarca** [comment]
![Cuplikan layar 2024-01-04 131859](https://github.com/ianka/xfs_undelete/assets/17668827/96a2e8af-aace-458a-8eda-4e3fa4bf41c0)

**Comment by Rendiarca** [comment]
On Centos 7 Cannot install Tcl >= 8.6

**Comment by ianka** [comment]
> On Centos 7 Cannot install Tcl >= 8.6

I see. You could define an equivalent of `lmap` inside the xfs_undelete script. Add the following lines between line 11 and line 12.

```
proc lmap {_var list body} {
    upvar 1 $_var var
    set res {}
    foreach var $list {lappend res [uplevel 1 $body]}
    set res
}
```

You will likely run into the next problem. Don't  give up. I try to help you with that one, too.

**Comment by Rendiarca** [comment]
Thank you so much. u save me a lot. but recovery got
![Cuplikan layar 2024-01-05 113254](https://github.com/ianka/xfs_undelete/assets/17668827/6e319305-79dc-4f1e-807f-141c2a85d62a)
 one huge pdf file. any advice please?

**Comment by Rendiarca** [comment]
Thank you so much. u save me a lot. but recovery got one huge pdf file. any advice please?

![Cuplikan layar 2024-01-05 113254](https://github.com/ianka/xfs_undelete/assets/17668827/5b1d350f-a27b-419c-ba73-4f5a6d49a3b4)

**Comment by ianka** [comment]
Those files are likely bogus. Xfs_undelete makes some guesses which inode records contain invalid data and does not recover those but it may guess wrong. You can delete those huge files from your output directory so you have again enough space on it, then continue the recovery with the additional option `-s 558929474`, as xfs_undelete suggested it.

**Comment by ianka** [comment]
Ahh, wait … hmm. That's exactly the inode it wants to restart at.

Please look for the following snippet, should be around line 172 in your current script:

		## Skip if we haven't passed the start inode yet.
		if {!$::passedstart} continue

After that line, insert as a patch the following line:

		if {$inode eq "558929474"} continue

That makes xfs_undelete skip that stupid inode. I will later add a permanent solution for such problems to the script.

**Comment by ianka** [comment]
I've just released [v13.0](https://github.com/ianka/xfs_undelete/releases/tag/v13.0) which adds that lmap snippet for Tcl-8.5 compatibility, adds the -x option to exclude a list of inode numbers from recovery, and adds the -S option to exclude files larger than the specified size from recovery. With the latter two, you can easily exclude those giant bogus files.

**Comment by ianka** [comment]
You haven't commented on it but I think this problem is solved from [v13.1](https://github.com/ianka/xfs_undelete/releases/tag/v13.1). I close this issue. Feel free to reopen if the problem isn't solved for you.
